File: caraml-console-android/src/main/java/org/liquidplayer/caraml/console/ConsoleJS.java
/*
 * Copyright (c) 2019 Eric Lange
 *
 * Distributed under the MIT License.  See LICENSE.md at
 * https://github.com/LiquidPlayer/caraml-console for terms and conditions.
 */
package org.liquidplayer.caraml.console;

import android.content.Context;
import android.graphics.Color;
import android.os.Handler;
import android.os.Looper;
import android.view.View;

import org.liquidplayer.caraml.CaramlJS;
import org.liquidplayer.caraml.CaramlSurface;
import org.liquidplayer.javascript.JSContext;
import org.liquidplayer.javascript.JSException;
import org.liquidplayer.javascript.JSFunction;
import org.liquidplayer.javascript.JSObject;
import org.liquidplayer.javascript.JSValue;

import java.util.HashMap;
import java.util.UUID;

public class ConsoleJS extends JSObject implements CaramlSurface, JSContext.IJSExceptionHandler {
    public ConsoleJS(final Context androidContext, final JSContext context, final JSValue opts) {
        super(context);
        currentState = State.Init;
        JSFunction createEmitter = new JSFunction(context, "createEmitter",
                "let e=new events(); Object.assign(thiz,e); " +
                "thiz.__proto__=e.__proto__", "thiz");
        createEmitter.call(null, this);
        emit = property("emit").toFunction();
        attachPromise = null;
        detachPromise = null;

        uuid = UUID.randomUUID().toString();
        sessionMap.put(uuid, this);

        // defaults
        this.backgroundColor = Color.BLACK;
        this.textColor = Color.GREEN;
        this.fontSize = 12.0f;

        if (opts != null && opts.isObject()) {
            JSObject options = opts.toObject();
            JSValue backgroundColor = options.property("backgroundColor");
            if (backgroundColor.isString())
                this.backgroundColor = Color.parseColor(backgroundColor.toString());
            JSValue textColor = options.property("textColor");
            if (textColor.isString())
                this.textColor = Color.parseColor(textColor.toString());
            JSValue fontSize = options.property("fontSize");
            if (fontSize.isNumber()) {
                this.fontSize = fontSize.toNumber().floatValue();
            }
        }

        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                currentView = new ConsoleSurface(androidContext);
                currentView.setSession(ConsoleJS.this);
                currentState = State.Detached;
                context.getGroup().schedule(new Runnable() {
                    @Override
                    public void run() {
                        emit.call(ConsoleJS.this, "ready");
                    }
                });
            }
        });
    }

    /*--
    /* JavaScript API
    /*--*/

    @jsexport @SuppressWarnings("unused")
    JSObject attach(JSValue value) {
        attachPromise = getContext().evaluateScript(createPromiseObject).toObject();
        try {
            if (value == null)
                throw new RuntimeException("attach: first argument must be a caraml object");
            if (currentState != State.Detached)
                throw new RuntimeException("attach: must be in detached state");
            caramlJS = CaramlJS.from(value);
            currentState = State.Attaching;
            caramlJS.attach(this);
        } catch (RuntimeException e) {
            onError(e);
        }

        return attachPromise.property("promise").toObject();
    }

    @jsexport @SuppressWarnings("unused")
    String state() {
        switch (currentState) {
            case Init:      return "init";
            case Attaching: return "attaching";
            case Attached:  return "attached";
            case Detaching: return "detaching";
            case Detached:  return "detached";
        }
        if (BuildConfig.DEBUG) throw new AssertionError();
        return "error";
    }

    @jsexport @SuppressWarnings("unused")
    JSObject detach() {
        detachPromise = getContext().evaluateScript(createPromiseObject).toObject();
        JSObject promise = detachPromise.property("promise").toObject();
        if (currentState == State.Detached) {
            emit.call(this, "detached");
            detachPromise.property("resolve").toFunction().call(null);
            detachPromise = null;
        } else if (currentState == State.Attached) {
            if (BuildConfig.DEBUG && caramlJS == null) throw new AssertionError();
            currentState = State.Detaching;
            caramlJS.detach();
        } else {
            detachPromise.property("reject").toFunction().
                    call(null, "Attach/detach pending");
            detachPromise = null;
        }
        return promise;
    }

    @jsexport @SuppressWarnings("unused")
    void write(String out) {
        if (currentView != null && out != null) currentView.print(out);
    }

    /*--
    /* CaramlSurface implementation
    /*--*/

    @Override
    public View getView() {
        return currentView;
    }

    @Override
    public void onAttached(boolean fromRestore) {
        currentState = State.Attached;
        getContext().setExceptionHandler(this);
        console_log = getContext().property("console").toObject().property("log").toFunction();
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                if (currentView != null) {
                    currentView.inputBox.setEnabled(true);
                }
            }
        });

        if (!fromRestore) {
            if (attachPromise != null) {
                attachPromise.property("resolve").toFunction().call(null);
            }
            emit.call(this, "attached");
            attachPromise = null;
        }
    }

    @Override
    public void onDetached() {
        currentState = State.Detached;

        console_log = null;
        if (uuid != null) {
            sessionMap.remove(uuid);
        }

        if (currentView != null) currentView.detach();
        caramlJS = null;

        emit.call(this, "detached");
        if (detachPromise != null) {
            detachPromise.property("resolve").toFunction().call(null);
        }
        detachPromise = null;
    }

    @Override
    public void onError(Exception e) {
        if (currentState != State.Init) {
            currentState = State.Detached;
        }
        JSObject promise;
        if (attachPromise != null) promise = attachPromise;
        else promise = detachPromise;

        if (promise != null) {
            promise.property("reject").toFunction().call(null, e.getMessage());
        }
        emit.call(this, "error", e.getMessage());
        attachPromise = detachPromise = null;
    }

    /*--
    /* Console session
    /*--*/

    static ConsoleJS getSessionFromUUID(String id) {
        return sessionMap.get(id);
    }

    String getSessionUUID() { return uuid; }

    void setCurrentView(ConsoleSurface view) {
        currentView = view;
    }

    int getBackgroundColor() {
        return backgroundColor;
    }

    int getTextColor() {
        return textColor;
    }

    float getFontSize() {
        return fontSize;
    }

    void removeCurrentView(ConsoleSurface view) {
        if (currentView == view) currentView = null;
    }

    void processCommand(final String cmd) {
        processedException = false;
        new Thread(new Runnable() {
            @Override
            public void run() {
                JSValue output = getContext().evaluateScript(cmd);
                if (!processedException && console_log != null) {
                    console_log.call(null, output);
                }
            }
        }).start();
    }

    void resize(final int columns, final int rows) {
        getContext().getGroup().schedule(new Runnable() {
            @Override
            public void run() {
                JSValue resize = property("resize");
                if (resize.isObject() && resize.toObject().isFunction()) {
                    resize.toFunction().call(null, columns, rows);
                }
            }
        });
    }

    /*--
    /* IJSExceptionHandler implementation
    /*--*/

    @Override
    public void handle(final JSException e) {
        processedException = true;
        if (currentView != null) {
            currentView.println("\u001b[31m" + e.stack());
        }
    }

    /*--
    /* private statics
    /*--*/

    private static final String createPromiseObject =
        "(()=>{" +
        "  var po = {}; var clock = true;" +
        "  var timer = setInterval(()=>{if(!clock) clearTimeout(timer);}, 100); "+
        "  po.promise = new Promise((resolve,reject)=>{po.resolve=resolve;po.reject=reject});" +
        "  po.promise.then(()=>{clock=false}).catch(()=>{clock=false});" +
        "  return po;" +
        "})();";
    private static HashMap<String,ConsoleJS> sessionMap = new HashMap<>();

    /*--
    /* session privates
    /*--*/

    private final JSFunction emit;
    private final String uuid;
    private JSObject attachPromise;
    private JSObject detachPromise;
    private ConsoleSurface currentView;
    private JSFunction console_log = null;
    private boolean processedException = false;
    private CaramlJS caramlJS;
    private int backgroundColor;
    private int textColor;
    private float fontSize;

    private enum State {
        Init,
        Attaching,
        Attached,
        Detaching,
        Detached,
    }
    private State currentState;
}

File: caraml-console/caraml-console/ConsoleJS.m
//
//  ConsoleJS.m
//  caraml-console
//
//  Created by Eric Lange on 1/10/19.
//  Copyright © 2019 LiquidPlayer. All rights reserved.
//

#import "ConsoleJS.h"
#import "UIColor+String.h"
#import "ConsoleView.h"
#import <caraml_core/caraml_core.h>

typedef enum _State {
    Init,
    Attaching,
    Attached,
    Detaching,
    Detached
} State;

@interface ConsoleJS()
@property (nonatomic, strong) JSValue* thiz;
@property (nonatomic, strong, readonly) JSValue* emitFunc;
@end

@implementation ConsoleJS {
    State currentState_;
    JSValue* attachPromise_;
    JSValue* detachPromise_;
    ConsoleView* currentView_;
    BOOL processedException_;
    LCProcess *process_;
    JSContext *context_;
    JSValue *emitFunc_;
    LCCaramlJS *caramlJS_;
}

static NSString* createJSS =
    @"((thiz) => {"
    @"  let e=new events();"
    @"  e.attach = thiz.attach;"
    @"  e.detach = thiz.detach;"
    @"  e.state  = thiz.state;"
    @"  e.write  = thiz.write;"
    @"  return e;"
    @"})";

static NSString* createPromiseObjectS =
    @"(()=>{"
    @"  var po = {}; var clock = true;"
    @"  var timer = setInterval(()=>{if(!clock) clearTimeout(timer);}, 100); "
    @"  po.promise = new Promise((resolve,reject)=>{po.resolve=resolve;po.reject=reject});"
    @"  po.promise.then(()=>{clock=false}).catch(()=>{clock=false});"
    @"  return po;"
    @"})()";

static NSString* emitFuncS = @"() => arguments[0].emit.call(...arguments)";

- (id) init:(JSContext*)context service:(LCMicroService*)service opts:(JSValue*)opts
{
    self = [super init];
    if (self != nil) {
        currentState_ = Init;
        process_ = service.process;
        context_ = context;
        
        // defaults
        _backgroundColor = UIColor.blackColor;
        _textColor = UIColor.greenColor;
        _fontSize = @(12);
        
        if (opts != nil && [opts isObject]) {
            JSValue* bg = opts[@"backgroundColor"];
            JSValue* tc = opts[@"textColor"];
            JSValue* fs = opts[@"fontSize"];
            if ([bg isString]) {
                UIColor *bgc = [UIColor colorWithString:[bg toString]];
                if (bgc != nil) _backgroundColor = bgc;
            }
            if ([tc isString]) {
                UIColor *tcc = [UIColor colorWithString:[bg toString]];
                if (tcc != nil) _textColor = tcc;
            }
            if ([fs isNumber]) {
                _fontSize = [fs toNumber];
            }
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            self->currentView_ = [[ConsoleView alloc] initWithFrame:CGRectMake(0, 0, 0, 0)];
            [self->currentView_ setSession:self];
            self->currentState_ = Detached;
            [self->process_ async:^(JSContext *context) {
                [self.emitFunc callWithArguments:@[self.thiz, @"ready"]];
            }];
        });
    }
    return self;
}

- (JSValue *)emitFunc
{
    if (emitFunc_ == nil) {
        emitFunc_ = [context_ evaluateScript:emitFuncS];
    }
    assert(emitFunc_ != nil);
    return emitFunc_;
}

+ (JSValue *) createConsole:(JSContext*)context
                 service:(LCMicroService*)service
                    opts:(JSValue*)opts
{
    ConsoleJS *consoleJS = [[ConsoleJS alloc] init:context service:service opts:opts];
    JSValue *createJS = [context evaluateScript:createJSS];
    JSValue *nativeObject = [createJS callWithArguments:@[consoleJS]];
    consoleJS.thiz = nativeObject;
    return nativeObject;
}

- (void) resize:(int)columns rows:(int)rows
{
    [process_ async:^(JSContext *context) {
        [self.emitFunc callWithArguments:@[self.thiz, @"resize", @(columns), @(rows)]];
    }];
}

- (void) processCommand:(NSString*)cmd
{
    processedException_ = NO;
    if (process_ != nil) {
        [process_ async:^(JSContext *context) {
            JSValue *output = [context evaluateScript:cmd];
            if (!self->processedException_) {
                [context[@"console"] invokeMethod:@"log" withArguments:@[output]];
            }
        }];
    }
}

#pragma JavaScript API

- (JSValue*) attach:(JSValue*)value
{
    attachPromise_ = [context_ evaluateScript:createPromiseObjectS];
    @try {
        if (value == nil || !value.isObject)
            @throw [NSException exceptionWithName:@"attach: first argument must be a caraml object"
                                           reason:nil userInfo:nil];
        if (currentState_ != Detached)
            @throw [NSException exceptionWithName:@"attach: must be in detached state"
                                           reason:nil userInfo:nil];
        caramlJS_ = [LCCaramlJS from:value];
        currentState_ = Attaching;
        [caramlJS_ attach:self];
    } @catch (NSException* e) {
        [self onError:e];
    }
    return attachPromise_[@"promise"];
}

- (NSString*) state
{
    switch (currentState_) {
        case Init:      return @"init";
        case Attaching: return @"attaching";
        case Attached:  return @"attached";
        case Detaching: return @"detaching";
        case Detached:  return @"detached";
    }
    assert(0);
}

- (JSValue *) detach
{
    detachPromise_ = [context_ evaluateScript:createPromiseObjectS];
    JSValue *promise = detachPromise_[@"promise"];
    if (currentState_ == Detached) {
        [self.emitFunc callWithArguments:@[self.thiz, @"detached"]];
        [detachPromise_[@"resolve"] callWithArguments:@[]];
        detachPromise_ = nil;
    } else if (currentState_ == Attached) {
        assert(caramlJS_ != nil);
        currentState_ = Detaching;
        [caramlJS_ detach];
    } else {
        [detachPromise_[@"reject"] callWithArguments:@[@"Attach/detach pending"]];
        detachPromise_ = nil;
    }
    return promise;
}

- (void) write:(NSString*) out
{
    if (currentView_ != nil && out != nil) {
        [currentView_.console print:out];
    }
}

#pragma LCCaramlSurface

- (UIView*) getView
{
    return currentView_;
}

- (void) onAttached:(BOOL)fromRestore
{
    currentState_ = Attached;
    
    [process_ async:^(JSContext *context) {
        context.exceptionHandler = ^(JSContext* context, JSValue *exception) {
            self->processedException_ = YES;
            [context[@"console"] invokeMethod:@"error" withArguments:@[exception]];
        };
        if (!fromRestore) {
            if (self->attachPromise_ != nil) {
                [self->attachPromise_[@"resolve"] callWithArguments:@[]];
            }
            [self.emitFunc callWithArguments:@[self.thiz, @"attached"]];
            self->attachPromise_ = nil;
        }
    }];
}

- (void) onDetached
{
    currentState_ = Detached;
    if (currentView_ != nil) [currentView_ detach];
    caramlJS_ = nil;
    
    [process_ async:^(JSContext *context) {
        [self.emitFunc callWithArguments:@[self.thiz, @"detached"]];
        if (self->detachPromise_ != nil) {
            [self->detachPromise_[@"resolve"] callWithArguments:@[]];
        }
        self->detachPromise_ = nil;
    }];
}

- (void) onError:(NSException*) e
{
    if (currentState_ != Init) {
        currentState_ = Detached;
    }
    JSValue *promise = attachPromise_ ? attachPromise_ : detachPromise_;
    [process_ async:^(JSContext *context) {
        if (promise != nil) {
            [promise[@"reject"] callWithArguments:@[e.name]];
        }
        [self.emitFunc callWithArguments:@[self.thiz, @"error", e.name]];
    }];
    attachPromise_ = detachPromise_ = nil;
}

@end
